src/efftester.ml

```diff
@@ -99,14 +103,25 @@ type lit =
   | LitStr of string
   | LitList of lit list
 
+(** type term is used to represent all syntax constructs (of OCaml) available in our
+generator  *)
 type term =
   | Lit of lit
   | Variable of etype * variable
+  (* [Constructor (type, name, load)] is used to construct ADT variants *)
+  | Constructor of etype * string * term option
```

Comment:
I would expect `term list` rather than `term option` to represent the arguments in a more general way (for example, if later we decided to integrate lists as special cases of ADT, and use `Constructor` for the list-consing `foo :: bar`).

Reply:
I initially started off with exact same recommendations you gave regarding the representation of `Constructor of adt_constr * type * term list`, where `type adt_constr = {name : string; arity : int}`. However, I decided to move away from those for following reasons:

1) (most important reason) When we use `term list` instead of `term option`, we have to pattern match for `[]`, `[single_load]`, and wildcard. In my understanding, wildcard causes *redundancy* AND is useless in most ADTs, as variants *either have a load or do not*. 
(I hope I am using correct terminology or at least it's understandable) Pattern matching with options is easier to get exhaustive and more apt to represent presence of load. Feel free to correct me if I am wrong.

2) Again, I included the arity first, but then considered it unnecessarily create an additional invariant to hold (`arity = List.length load_lst`). Again, either there is load or there is not.

Reply:
Note: I am, however, open to changes if there are good reasons for us to move from an option to a list.

Reply:
Suppose later we want to generate values for a different algebraic datatype:

```ocaml
type foo = Foo of 'a * 'b * 'c
```

then we will naturally want to represent `Foo(t,u,v)` as `Constructor(, "Foo", [t; u; v])`, which doesn't fit your current definition.

Of course, having a list of arguments means that we can represent ill-typed terms like `Constructor(..., "None", [foo; bar])` or `Constructor(..., "Some", [])`, but that (the ability to represent ill-typed terms) was already the case with `option`.

You do lose the ability to do exhaustive pattern-matching on the arguments, but that comes from the ability the represent arbitrary number of constructor parameters, and I think the latter is more important in that case.